[7x]gprecoverseg: Raise exception when -F option is used in rebalance and recovery to new host options

piyushc01 opened the pull request:
**Issue:**
Currently, gprecoverseg performs rebalance operation (-r option) which is an incremental operation. To recover segments to a new host (-p option) operation is always a full recovery operation. 
Providing the `-F `flag along with the` -r `or -p option makes no sense. 

**Solution:**
To raise the exception when the Full Recovery option (-F)  is specified along with:
1. Recovery to new host (-p option)
2. Rebalance Operation (-r option)

Changes in this PR: 

- Raise error when -F option is passed with -r or -p
- Moved gprecoverseg flags validation to a separate function. 
- Added tests to verify rebalance and new host options.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

Annu149 commented on gpMgmt/bin/gppylib/programs/clsRecoverSegment.py:
```diff
@@ -386,6 +357,45 @@ def signal_handler(sig, frame):
+            if self.__options.rebalanceSegments:
+                logger.warn("-F option not supported with -r option. Ignoring -F flag and continuing with recovery.")
+            if self.__options.newRecoverHosts is not None:
+                logger.warn("-F option not supported with -p option. Ignoring -F flag and continuing with recovery.")
```
it should be 'continuing with rebalance' right?

piyushc01 replied:
rebalance is also a recovery operation hence added generic recovery text.
@nimish350  can you review the warning to be printed?

nimish350 replied:
Should be "option is not supported" instead of "option not supported'. Rest looks good. 👍

piyushc01 replied:
Updated

Annu149 commented on gpMgmt/test/behave/mgmt_utils/gprecoverseg_newhost.feature:
```diff
@@ -221,6 +221,32 @@ Feature: gprecoverseg tests involving migrating to a new host
       And the "before" and "after_recreation" cluster configuration matches with the expected for gprecoverseg newhost
 
+   @concourse_cluster
+    Scenario: Recovery to new host with full recovery option ignores -F flag and prints warning
```
Do we really need separate scenario for this warning message? Can't we just add this warning part to existing scenarios?

piyushc01 replied:
This is a copy of the existing scenario of recovery to a new host with -F option.
Existing scenarios are not checking for any error/warning messages.

Annu149 commented on gpMgmt/bin/gppylib/programs/clsRecoverSegment.py:
```diff
@@ -386,6 +357,45 @@ def signal_handler(sig, frame):
+            raise ProgramArgumentValidationException(
+                "Invalid parallelPerHost value provided with -b argument: %d" % self.__options.parallelPerHost)
+        self.__pool = WorkerPool(self.__options.parallelDegree)
+        gpEnv = GpCoordinatorEnvironment(self.__options.coordinatorDataDirectory, True)
```
I think this function should only validate options and return nothing. Getting the gpEnv can be done in the main code as it was done previously

piyushc01 replied:
Updated

nimish350 commented:
@piyushc01 @shrakesh I tested this out in my environment. +1 for the -F and -p combination warning. 

For rebalance though, I feel we should abort with an ERROR. The reasoning is if the customer wants to do "gprecoverseg -F" and inadvertently types "gprecoverseg -F -r", it might proceed to do a rebalance of any segments which are not in the preferred role. This might be a bad thing if the customer never intended to do a rebalance as it might cancel any ongoing jobs / SQLs. On a production environment, this might be a bad thing. 

Let me know your opinion.

shrakesh commented on gpMgmt/test/behave/mgmt_utils/gprecoverseg_newhost.feature:
```diff
@@ -238,3 +264,4 @@ Feature: gprecoverseg tests involving migrating to a new host
          When the user runs gprecoverseg with input file and additional args "-av"
          Then gprecoverseg should return a return code of 2
          And gprecoverseg  should print "The recovery target segment sdw5 (content 0) is unreachable" escaped to stdout
+      
```
remove no empty line warning.

shrakesh commented:
> @piyushc01 @shrakesh I tested this out in my environment. +1 for the -F and -p combination warning.
> 
> For rebalance though, I feel we should abort with an ERROR. The reasoning is if the customer wants to do "gprecoverseg -F" and inadvertently types "gprecoverseg -F -r", it might proceed to do a rebalance of any segments which are not in the preferred role. This might be a bad thing if the customer never intended to do a rebalance as it might cancel any ongoing jobs / SQLs. On a production environment, this might be a bad thing.
> 
> Let me know your opinion.

@nimish350 I agree with you and even if with -p for differential recovery we are throwing out validation exception(error). 
IMO it should be consistent with all recovery-type options.  There is no sense for the user to give -p -F together, 
I feel -p -F the expectation of the user could be possible to have it recovered it new host, so it better to error out. 

differential option with -p
```
gpdb git:(diff_performance) ✗ gprecoverseg -p localhost --differential
20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-Starting gprecoverseg with args: -p localhost --differential
20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-local Greenplum Version: 'postgres (Greenplum Database) 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev'
20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-coordinator Greenplum Version: 'PostgreSQL 12.12 (Greenplum Database 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev) on x86_64-apple-darwin22.6.0, compiled by Apple clang version 14.0.3 (clang-1403.0.22.14.1), 64-bit compiled on Aug 22 2023 23:42:20 (with assert checking) Bhuvnesh C.'
20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[ERROR]:-gprecoverseg: error: Only one of -i, -p, -r and --differential may be specified
```

differential option with -r.
```
gpdb git:(diff_performance) ✗ gprecoverseg -r --differential
20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-Starting gprecoverseg with args: -r --differential
20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-local Greenplum Version: 'postgres (Greenplum Database) 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev'
20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-coordinator Greenplum Version: 'PostgreSQL 12.12 (Greenplum Database 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev) on x86_64-apple-darwin22.6.0, compiled by Apple clang version 14.0.3 (clang-1403.0.22.14.1), 64-bit compiled on Aug 22 2023 23:42:20 (with assert checking) Bhuvnesh C.'
20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[ERROR]:-gprecoverseg: error: Only one of -i, -p, -r and --differential may be specified
```

nimish350 commented:
> > @piyushc01 @shrakesh I tested this out in my environment. +1 for the -F and -p combination warning.
> > For rebalance though, I feel we should abort with an ERROR. The reasoning is if the customer wants to do "gprecoverseg -F" and inadvertently types "gprecoverseg -F -r", it might proceed to do a rebalance of any segments which are not in the preferred role. This might be a bad thing if the customer never intended to do a rebalance as it might cancel any ongoing jobs / SQLs. On a production environment, this might be a bad thing.
> > Let me know your opinion.
> 
> @nimish350 I agree with you and even if with -p for differential recovery we are throwing out validation exception(error). IMO it should be consistent with all recovery-type options. There is no sense for the user to give -p -F together, I feel -p -F the expectation of the user could be possible to have it recovered it new host, so it better to error out.
> 
> differential option with -p
> 
> ```
> gpdb git:(diff_performance) ✗ gprecoverseg -p localhost --differential
> 20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-Starting gprecoverseg with args: -p localhost --differential
> 20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-local Greenplum Version: 'postgres (Greenplum Database) 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev'
> 20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-coordinator Greenplum Version: 'PostgreSQL 12.12 (Greenplum Database 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev) on x86_64-apple-darwin22.6.0, compiled by Apple clang version 14.0.3 (clang-1403.0.22.14.1), 64-bit compiled on Aug 22 2023 23:42:20 (with assert checking) Bhuvnesh C.'
> 20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[ERROR]:-gprecoverseg: error: Only one of -i, -p, -r and --differential may be specified
> ```
> 
> differential option with -r.
> 
> ```
> gpdb git:(diff_performance) ✗ gprecoverseg -r --differential
> 20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-Starting gprecoverseg with args: -r --differential
> 20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-local Greenplum Version: 'postgres (Greenplum Database) 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev'
> 20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-coordinator Greenplum Version: 'PostgreSQL 12.12 (Greenplum Database 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev) on x86_64-apple-darwin22.6.0, compiled by Apple clang version 14.0.3 (clang-1403.0.22.14.1), 64-bit compiled on Aug 22 2023 23:42:20 (with assert checking) Bhuvnesh C.'
> 20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[ERROR]:-gprecoverseg: error: Only one of -i, -p, -r and --differential may be specified
> ```



I agree. When I tested this out, the WARNING seems to get lost in the output. @piyushc01 Can you please change this to a ERROR and exit out.

piyushc01 commented:
Updated the PR with the bellow change. 
1. With -F and -r option, print error and exit 
2. Wirg -F and -p options, print a warning, and continue.
The pipeline is also green. 
I have updated the PR description also.

> > > @piyushc01 @shrakesh I tested this out in my environment. +1 for the -F and -p combination warning.
> > > For rebalance though, I feel we should abort with an ERROR. The reasoning is if the customer wants to do "gprecoverseg -F" and inadvertently types "gprecoverseg -F -r", it might proceed to do a rebalance of any segments which are not in the preferred role. This might be a bad thing if the customer never intended to do a rebalance as it might cancel any ongoing jobs / SQLs. On a production environment, this might be a bad thing.
> > > Let me know your opinion.
> > 
> > 
> > @nimish350 I agree with you and even if with -p for differential recovery we are throwing out validation exception(error). IMO it should be consistent with all recovery-type options. There is no sense for the user to give -p -F together, I feel -p -F the expectation of the user could be possible to have it recovered it new host, so it better to error out.
> > differential option with -p
> > ```
> > gpdb git:(diff_performance) ✗ gprecoverseg -p localhost --differential
> > 20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-Starting gprecoverseg with args: -p localhost --differential
> > 20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-local Greenplum Version: 'postgres (Greenplum Database) 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev'
> > 20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-coordinator Greenplum Version: 'PostgreSQL 12.12 (Greenplum Database 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev) on x86_64-apple-darwin22.6.0, compiled by Apple clang version 14.0.3 (clang-1403.0.22.14.1), 64-bit compiled on Aug 22 2023 23:42:20 (with assert checking) Bhuvnesh C.'
> > 20230901:11:28:52:057088 gprecoverseg:shrakeshSMD6M:shrakesh-[ERROR]:-gprecoverseg: error: Only one of -i, -p, -r and --differential may be specified
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > differential option with -r.
> > ```
> > gpdb git:(diff_performance) ✗ gprecoverseg -r --differential
> > 20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-Starting gprecoverseg with args: -r --differential
> > 20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-local Greenplum Version: 'postgres (Greenplum Database) 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev'
> > 20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[INFO]:-coordinator Greenplum Version: 'PostgreSQL 12.12 (Greenplum Database 7.0.0-beta.5+dev.85.gc3d97a3d441 build dev) on x86_64-apple-darwin22.6.0, compiled by Apple clang version 14.0.3 (clang-1403.0.22.14.1), 64-bit compiled on Aug 22 2023 23:42:20 (with assert checking) Bhuvnesh C.'
> > 20230901:11:28:18:056970 gprecoverseg:shrakeshSMD6M:shrakesh-[ERROR]:-gprecoverseg: error: Only one of -i, -p, -r and --differential may be specified
> > ```
> 
> I agree. When I tested this out, the WARNING seems to get lost in the output. @piyushc01 Can you please change this to a ERROR and exit out.

ShirishaRao commented on gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature:
```diff
@@ -471,6 +471,22 @@ Feature: gprecoverseg tests
+        And user stops all primary processes
+        And user can start transactions
+        And a gprecoverseg directory under '/tmp' with mode '0700' is created
+        And a gprecoverseg input file is created
```
As we are not checking for 'gprecoverseg -i' option, we can skip the input file creation here ?

piyushc01 replied:
That's a good point. 
Currently with -i option we need to explicitly mention the type of recovery (full/incremental). 
Even if -F is provided or not, we do the recovery as specified in the config file only.  
@nimish350 Do you think we should print a warning in case of the -F flag along with -i? 

Example, currently it ignores the -F flag: 
```
$ gprecoverseg -i test.conf -F
[INFO]:-Recovery 1 of 1
[INFO]:----------------------------------------------------------
[INFO]:-   Synchronization mode                 = Full
[INFO]:-   Failed instance host                    =  sdw1
[INFO]:-   Failed instance address              = sdw1
[INFO]:-   Failed instance directory            = /data/gpseg0
[INFO]:-   Failed instance port                 = 7002
[INFO]:-   Recovery Source instance host        = sdw1
[INFO]:-   Recovery Source instance address     = sdw2
[INFO]:-   Recovery Source instance directory   = /data/mirror/gpseg0
[INFO]:-   Recovery Source instance port        = 7005
[INFO]:-   Recovery Target                      = in-place
[INFO]:----------------------------------------------------------

```

nimish350 replied:
> 

When specifying -i with -F, it tells the utility to only do a full recovery for the segments mentioned in the file. This is a helpful way to selectively provide the segments that you want to run full recovery on. My take is that we should not do any changes for the -i and -F combination.

nimish350 replied:
No warning needed for -i and -F as per me. This two flags have been used together since 4.3.x.

nimish350 commented:
@piyushc01 Tested, LGTM +1

```
20230913:09:53:49:032150 gprecoverseg:mdwHMD6R:mdw-[INFO]:-Starting gprecoverseg with args: -F -r
20230913:09:53:49:032150 gprecoverseg:mdwHMD6R:mdw-[ERROR]:-gprecoverseg: error: -F option is not supported with -r option
```

Annu149 reviewed (approved):
LGTM

nimish350 reviewed (approved):
LGTM +1

shrakesh reviewed (approved):
LGTM

ShirishaRao reviewed (approved):
LGTM!
